src/main/java/com/metamx/emitter/core/HttpPostEmitter.java

```diff
@@ -166,315 +191,401 @@ public void emit(Event event)
-        lastBufferFullWarning = now;
-        messagesDroppedSinceLastBufferFullWarning = 0;
+  /**
+   * Called from {@link Batch} only once for each Batch in existence.
+   */
+  void onSealExclusive(Batch batch)
+  {
+    buffersToEmit.add(batch);
+    wakeUpEmittingThread();
+    if (!isTerminated()) {
+      int nextBatchNumber = EmittedBatchCounter.nextBatchNumber(batch.batchNumber);
+      if (!concurrentBatch.compareAndSet(batch, new Batch(this, acquireBuffer(), nextBatchNumber))) {
+        // If compareAndSet failed, the service is closed concurrently.
+        Preconditions.checkState(isTerminated());
       }
     }
   }
 
+  private void wakeUpEmittingThread()
+  {
+    LockSupport.unpark(emittingThread);
+  }
+
   @Override
   public void flush() throws IOException
   {
-    if (started.get()) {
-      final Future future = exec.submit(new EmittingRunnable(version.get()));
+    awaitStarted();
+    flush(concurrentBatch.get());
+  }
 
-      try {
-        future.get(config.getFlushTimeOut(), TimeUnit.MILLISECONDS);
-      }
-      catch (InterruptedException e) {
-        log.debug("Thread Interrupted");
-        Thread.currentThread().interrupt();
-        throw new IOException("Thread Interrupted while flushing", e);
-      }
-      catch (ExecutionException e) {
-        throw new IOException("Exception while flushing", e);
-      }
-      catch (TimeoutException e) {
-        throw new IOException(String.format("Timed out after [%d] millis during flushing", config.getFlushTimeOut()), e);
-      }
+  private void flush(Batch batch) throws IOException
+  {
+    if (batch == null) {
+      return;
+    }
+    batch.seal();
+    try {
+      emittedBatchCounter.awaitBatchEmitted(batch.batchNumber, config.getFlushTimeOut(), TimeUnit.MILLISECONDS);
+    }
+    catch (TimeoutException e) {
+      String message = String.format("Timed out after [%d] millis during flushing", config.getFlushTimeOut());
+      throw new IOException(message, e);
+    }
+    catch (InterruptedException e) {
+      log.debug("Thread Interrupted");
+      Thread.currentThread().interrupt();
+      throw new IOException("Thread Interrupted while flushing", e);
     }
   }
 
   @Override
   @LifecycleStop
   public void close() throws IOException
   {
-    synchronized (started) {
-      // flush() doesn't do things if it is not started, so flush must happen before we mark it as not started.
-      flush();
-      started.set(false);
-      exec.shutdown();
+    synchronized (startLock) {
+      if (running) {
+        running = false;
+        Batch lastBatch = concurrentBatch.getAndSet(null);
+        flush(lastBatch);
+        emittingThread.shuttingDown = true;
+        // Emitter is interrupted after the last batch is flushed.
+        emittingThread.interrupt();
+      }
     }
   }
 
-  private class EmittingRunnable implements Runnable
+  private class EmittingThread extends Thread
```

Comment:
@drcrallen emitting used to be done via scheduled tasks, submitted to a single-thread executor service. The logic is changed and now there are no tasks, just a separate thread that does some work and goes to sleep sometimes. So it could still be a runnable which runs forever in a single-thread executor service, that makes no sense to me.